app/src/main/java/org/mozilla/focus/utils/CoroutineScope.kt

```diff
@@ -0,0 +1,69 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.focus.utils
+
+import android.arch.lifecycle.Lifecycle
+import android.arch.lifecycle.LifecycleObserver
+import android.arch.lifecycle.OnLifecycleEvent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A [CoroutineScope] that is scoped to an Activity's lifecycle and launches jobs on the main thread.
+ *
+ * Callers must call [init] to register the scope to the Activity.
+ */
+class ActivityUiCoroutineScope : LifecycleObserver, CoroutineScope {
+
+    override val coroutineContext: CoroutineContext
+        get() {
+            if (!wasInitCalled) throw IllegalStateException("Expected init to be called before accessing context")
+            return Dispatchers.Main + lifecycleCancelJob
+        }
+
+    private var wasInitCalled = false
+    private val lifecycleCancelJob = Job()
+
+    fun init(lifecycle: Lifecycle) {
+        wasInitCalled = true
+        lifecycle.addObserver(this)
+    }
+
+    @OnLifecycleEvent(Lifecycle.Event.ON_DESTROY)
+    fun onDestroy() {
+        lifecycleCancelJob.cancel()
```

Comment:
It's less that I disagree, and more that I'm constantly wary of framework lifecycles. I don't think it's unreasonable to program defensively around them, but you're probably right.

Reply:
My concerns with programming defensively around them are:
- it's more (potentially buggy) code and increases complexity
- if what you're doing isn't documented and consistent, it will be confusing to new (or existing!) devs
- "programming defensively" is amorphous: which non-contract cases are you trying to protect against? `onDestroy` called twice? Activity re-use after destroy? In theory, the problem space is nearly infinite, it's impossible to protect against all of it, and you may run into issues with devs trying to protect against different things

In theory, they're a contract so other devs should share an understanding of how they work and you should be able to rely on them: you can use assertions to make sure your assumptions about the lifecycle are correct and make sure you're implementation is correct.

However, there are sometimes bugs in framework implementation (e.g. onStart called twice before onStop): tbh, I don't know how to handle these in a systematic way and fixing them one off is not a great solution either.

I think the only way defensive programming works really well against these lifecycle issues is if you're able to mostly abstract the lifecycle away (e.g. Android architecture components) or greatly simplify it and you consistently use it. Hopefully, these solutions have taken into account the known bugs in lifecycle implementation (e.g. maybe make `onStart` called twice before `onStop` an idempotent operation).